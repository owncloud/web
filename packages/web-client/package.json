{
  "name": "@ownclouders/web-client",
  "version": "11.3.1",
  "description": "ownCloud web client",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-client",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-client"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "default": "./dist/web-client.js",
        "require": "./dist/web-client.cjs",
        "types": "./dist/src/index.d.ts"
      },
      "./graph": {
        "default": "./dist/web-client/graph.js",
        "require": "./dist/web-client/graph.cjs",
        "types": "./dist/src/graph/index.d.ts"
      },
      "./graph/generated": {
        "default": "./dist/web-client/graph/generated.js",
        "require": "./dist/web-client/graph/generated.cjs",
        "types": "./dist/src/graph/generated/index.d.ts"
      },
      "./ocs": {
        "default": "./dist/web-client/ocs.js",
        "require": "./dist/web-client/ocs.cjs",
        "types": "./dist/src/ocs/index.d.ts"
      },
      "./sse": {
        "default": "./dist/web-client/sse.js",
        "require": "./dist/web-client/sse.cjs",
        "types": "./dist/src/sse/index.d.ts"
      },
      "./webdav": {
        "default": "./dist/web-client/webdav.js",
        "require": "./dist/web-client/webdav.cjs",
        "types": "./dist/src/webdav/index.d.ts"
      }
    }
  },
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./graph": {
      "default": "./src/graph/index.ts"
    },
    "./graph/generated": {
      "default": "./src/graph/generated/index.ts"
    },
    "./ocs": {
      "default": "./src/ocs/index.ts"
    },
    "./sse": {
      "default": "./src/sse/index.ts"
    },
    "./webdav": {
      "default": "./src/webdav/index.ts"
    },
    "./l10n": {
      "default": "./l10n/translations.json"
    }
  },
  "scripts": {
    "generate-openapi": "rm -rf src/graph/generated && docker run --rm -v \"${PWD}/src/graph:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v1.0.yaml  -g typescript-axios -o /local/generated",
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && rm -rf package/dist/tests && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@microsoft/fetch-event-source": "^2.0.1",
    "@ownclouders/web-pkg": "workspace:*",
    "axios": "^1.7.7",
    "fast-xml-parser": "^4.5.0",
    "lodash-es": "^4.17.21",
    "luxon": "^3.5.0",
    "uuid": "^11.0.0",
    "webdav": "^5.7.1",
    "xml-js": "^1.6.11",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@types/luxon": "3.4.2",
    "@vitejs/plugin-vue": "5.1.4",
    "clean-publish": "5.0.0",
    "vite-plugin-dts": "4.2.4",
    "vite-plugin-node-polyfills": "0.22.0",
    "vite": "5.4.8"
  }
}
