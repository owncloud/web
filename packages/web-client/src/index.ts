import { graph } from './graph'
import { ocs } from './ocs'
import { webdav } from './webdav'

export * from './errors'
export * from './helpers'
export * from './utils'
export * from './constants'

export { graph, ocs, webdav }
