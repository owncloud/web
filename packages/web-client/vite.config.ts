import { join, resolve } from 'path'
import { defineConfig, searchForWorkspaceRoot } from 'vite'
import dts from 'vite-plugin-dts'
import { nodePolyfills } from 'vite-plugin-node-polyfills'

const projectRootDir = searchForWorkspaceRoot(process.cwd())

export default defineConfig({
  resolve: {
    alias: {
      crypto: join(projectRootDir, 'polyfills/crypto.js')
    }
  },
  build: {
    lib: {
      entry: {
        'web-client': resolve(__dirname, 'src/index.ts'),
        'web-client/graph': resolve(__dirname, 'src/graph/index.ts'),
        'web-client/graph/generated': resolve(__dirname, 'src/graph/generated/index.ts'),
        'web-client/ocs': resolve(__dirname, 'src/ocs/index.ts'),
        'web-client/sse': resolve(__dirname, 'src/sse/index.ts'),
        'web-client/webdav': resolve(__dirname, 'src/webdav/index.ts')
      }
    }
  },
  plugins: [
    nodePolyfills({
      exclude: ['crypto']
    }),
    dts({
      include: ['src'],
      outDir: 'dist/types',
      insertTypesEntry: true
    })
  ]
})
