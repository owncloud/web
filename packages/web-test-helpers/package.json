{
  "name": "@ownclouders/web-test-helpers",
  "version": "11.3.2",
  "description": "ownCloud web test helpers",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "main": "src/index.ts",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-test-helpers",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-test-helpers"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "import": "./dist/web-test-helpers.es.js",
        "require": "./dist/web-test-helpers.umd.cjs",
        "types": "./dist/src/index.d.ts"
      }
    }
  },
  "scripts": {
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "peerDependencies": {
    "@ownclouders/web-pkg": "workspace:^",
    "@vue/test-utils": "^2.4.6",
    "vue": "^3.5.10"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "5.1.4",
    "clean-publish": "^5.0.0",
    "vite-plugin-dts": "^4.2.3",
    "vite": "^5.4.8"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@casl/vue": "^2.2.2",
    "@ownclouders/design-system": "workspace:^",
    "@ownclouders/web-client": "workspace:^",
    "@pinia/testing": "^0.1.3",
    "axios": "1.7.7",
    "pinia": "2.2.6",
    "vitest-mock-extended": "2.0.2",
    "vue-router": "4.2.5",
    "vue3-gettext": "2.4.0"
  }
}
