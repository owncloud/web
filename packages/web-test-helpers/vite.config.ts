import { resolve } from 'path'
import { defineConfig } from 'vite'
import dts from 'vite-plugin-dts'
import pkg from './package.json' assert { type: 'json' }
import vue from '@vitejs/plugin-vue'

export default defineConfig({
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-test-helpers',
      fileName: (format) => `web-test-helpers.${format}.js`
    },
    rollupOptions: {
      external: [...Object.keys(pkg.dependencies), ...Object.keys(pkg.peerDependencies)]
    }
  },
  plugins: [vue(), dts({ include: ['src'], outDir: 'dist/types', insertTypesEntry: true })]
})
