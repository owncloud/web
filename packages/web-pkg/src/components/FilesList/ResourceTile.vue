<template>
  <div
    ref="observerTarget"
    class="oc-tile-card oc-card oc-card-default oc-rounded"
    :data-item-id="resource.id"
    :class="{
      'oc-tile-card-selected': isResourceSelected,
      'oc-tile-card-disabled': isResourceDisabled && !isProjectSpaceResource(resource),
      'state-trashed': isResourceDisabled && isProjectSpaceResource(resource)
    }"
    @contextmenu="$emit('contextmenu', $event)"
  >
    <div v-if="isHidden" class="oc-tile-card-lazy-shimmer"></div>
    <template v-else>
      <resource-link
        class="oc-card-media-top oc-flex oc-flex-center oc-flex-middle oc-m-rm"
        :resource="resource"
        :link="resourceRoute"
        :is-resource-clickable="isResourceClickable"
        tabindex="-1"
        @click="$emit('click')"
      >
        <div class="oc-tile-card-selection">
          <slot name="selection" :item="resource" />
        </div>
        <oc-tag
          v-if="isResourceDisabled && isProjectSpaceResource(resource)"
          class="resource-disabled-indicator oc-position-absolute"
          type="span"
        >
          <span v-text="$gettext('Disabled')" />
        </oc-tag>
        <div
          v-oc-tooltip="tooltipLabelIcon"
          class="oc-tile-card-preview oc-flex oc-flex-middle oc-flex-center"
          :aria-label="tooltipLabelIcon"
        >
          <div class="oc-tile-card-hover"></div>
          <slot name="imageField" :item="resource">
            <oc-img
              v-if="shouldDisplayThumbnails(resource)"
              class="tile-preview"
              :src="resource.thumbnail"
            />
            <resource-icon
              v-else
              :resource="resource"
              :size="resourceIconSize"
              class="tile-default-image oc-pt-xs"
            >
              <template v-if="showStatusIcon" #status>
                <oc-icon v-bind="statusIconAttrs" size="xsmall" />
              </template>
            </resource-icon>
          </slot>
        </div>
      </resource-link>
      <div class="oc-card-body oc-p-s">
        <div class="oc-flex oc-flex-between oc-flex-middle">
          <div class="oc-flex oc-flex-middle oc-text-truncate resource-name-wrapper">
            <resource-list-item
              :resource="resource"
              :is-icon-displayed="false"
              :is-extension-displayed="isExtensionDisplayed"
              :is-resource-clickable="isResourceClickable"
              :link="resourceRoute"
              @click="$emit('click')"
            />
          </div>
          <div class="oc-flex oc-flex-middle">
            <!-- Slot for indicators !-->
            <slot name="indicators" :item="resource" class="resource-indicators" />
            <!-- Slot for individual actions -->
            <slot name="actions" :item="resource" />
            <!-- Slot for contextmenu -->
            <slot name="contextMenu" :item="resource" />
          </div>
        </div>
        <p v-if="resourceDescription" class="oc-text-left oc-my-rm oc-text-truncate">
          <small v-text="resourceDescription" />
        </p>
      </div>
    </template>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, PropType } from 'vue'
import ResourceIcon from './ResourceIcon.vue'
import ResourceListItem from './ResourceListItem.vue'
import ResourceLink from './ResourceLink.vue'
import { isProjectSpaceResource, Resource } from '@ownclouders/web-client'
import { useGettext } from 'vue3-gettext'
import { isSpaceResource } from '@ownclouders/web-client'
import { isResourceTxtFileAlmostEmpty } from '../../helpers'
import { RouteLocationRaw } from 'vue-router'
import { useIsVisible } from '@ownclouders/design-system/src/composables'
import { customRef, ref, unref } from 'vue'

export default defineComponent({
  name: 'ResourceTile',
  components: { ResourceListItem, ResourceIcon, ResourceLink },
  props: {
    /**
     * Resource to be displayed within the tile
     */
    resource: {
      type: Object as PropType<Resource>,
      default: () => ({})
    },
    resourceRoute: {
      type: Object as PropType<RouteLocationRaw>,
      required: false,
      default: null
    },
    isResourceSelected: {
      type: Boolean,
      required: false,
      default: false
    },
    isResourceClickable: {
      type: Boolean,
      required: false,
      default: true
    },
    isResourceDisabled: {
      type: Boolean,
      required: false,
      default: false
    },
    isExtensionDisplayed: {
      type: Boolean,
      default: true
    },
    resourceIconSize: {
      type: String,
      default: 'xlarge',
      validator: (value: string) => {
        return ['large', 'xlarge', 'xxlarge', 'xxxlarge'].includes(value)
      }
    },
    lazy: {
      type: Boolean,
      default: false
    }
  },
  emits: ['click', 'contextmenu'],
  setup(props) {
    const { $gettext } = useGettext()

    const observerTarget = customRef((track, trigger) => {
      let $el: HTMLElement
      return {
        get() {
          track()
          return $el
        },
        set(value) {
          $el = value
          trigger()
        }
      }
    })

    const showStatusIcon = computed(() => {
      return props.resource.locked || props.resource.processing
    })

    const statusIconAttrs = computed(() => {
      if (props.resource.locked) {
        return {
          name: 'lock',
          fillType: 'fill'
        }
      }

      if (props.resource.processing) {
        return {
          name: 'loop-right',
          fillType: 'line'
        }
      }

      return {}
    })

    const tooltipLabelIcon = computed(() => {
      if (props.resource.locked) {
        return $gettext('This item is locked')
      }
      return null
    })
    const resourceDescription = computed(() => {
      if (isSpaceResource(props.resource)) {
        return props.resource.description
      }
      return ''
    })

    const shouldDisplayThumbnails = (resource: Resource) => {
      return resource.thumbnail && !isResourceTxtFileAlmostEmpty(resource)
    }

    const { isVisible } = props.lazy
      ? useIsVisible({
          target: observerTarget
        })
      : { isVisible: ref(true) }

    const isHidden = computed(() => !unref(isVisible))

    return {
      statusIconAttrs,
      showStatusIcon,
      tooltipLabelIcon,
      resourceDescription,
      shouldDisplayThumbnails,
      isProjectSpaceResource,
      isHidden,
      observerTarget
    }
  }
})
</script>

<style lang="scss">
.oc-tile-card {
  background-color: var(--oc-color-background-highlight) !important;
  box-shadow: none;
  height: 100%;
  display: flex;
  flex-flow: column;
  outline: 1px solid var(--oc-color-border);

  &-disabled {
    pointer-events: none;
    background-color: var(--oc-color-background-muted) !important;
    opacity: 0.7;
    filter: grayscale(0.6);

    // Show tooltip on status indicators without handler
    span.oc-status-indicators-indicator {
      pointer-events: all;
    }
  }

  &.state-trashed {
    .tile-image,
    .tile-default-image > svg {
      filter: grayscale(100%);
      opacity: 80%;
    }
  }

  .tile-default-image {
    position: relative;
  }

  .oc-card-media-top {
    position: relative;
    aspect-ratio: 16/9;
    justify-content: center;
    width: 100%;

    .oc-tag {
      color: var(--oc-color-text-default);

      &.resource-disabled-indicator {
        z-index: 1;
      }
    }

    .tile-preview {
      aspect-ratio: 16/9;
      height: 100%;
      object-fit: cover;
      width: 100%;
      border-top-left-radius: 5px;
      border-top-right-radius: 5px;
    }

    &:hover {
      .oc-tile-card-hover {
        opacity: 15%;
      }
    }
  }

  &-selected {
    outline: 2px solid var(--oc-color-swatch-primary-hover);

    .oc-tile-card-preview {
      width: calc(100% - var(--oc-space-medium));
      height: calc(100% - var(--oc-space-medium));

      .tile-preview,
      .oc-tile-card-hover {
        border-radius: 5px !important;
      }

      .oc-tile-card-hover {
        opacity: 10%;
      }
    }
  }

  &-selection {
    z-index: 1;
    position: absolute;
    top: 0;
    left: 0;

    input {
      background-color: var(--oc-color-background-muted);
    }

    input.oc-checkbox-checked {
      background-color: var(--oc-color-swatch-inverse-default);
    }
  }

  &-preview {
    position: absolute;
    height: 100%;
    width: 100%;
    text-align: center;
  }

  &-hover {
    position: absolute;
    width: 100%;
    height: 100%;
    background: #000;
    opacity: 0;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
  }

  .resource-name-wrapper {
    color: var(--oc-color-text-default);
    max-width: 70%;
    overflow: hidden;
  }

  &-lazy-shimmer {
    height: 120px;
    opacity: 0.2;
    position: relative;
    overflow: hidden;
  }

  &-lazy-shimmer::after {
    animation: shimmer 2s infinite;
    background-image: linear-gradient(
      90deg,
      rgba(#4c5f79, 0) 0,
      rgba(#4c5f79, 0.2) 20%,
      rgba(#4c5f79, 0.5) 60%,
      rgba(#4c5f79, 0)
    );
    bottom: 0;
    content: '';
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
  }

  @keyframes shimmer {
    100% {
      transform: translateX(100%);
    }
  }
}
</style>
