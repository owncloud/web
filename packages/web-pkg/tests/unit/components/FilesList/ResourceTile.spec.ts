import { defaultPlugins, shallowMount } from '@ownclouders/web-test-helpers'
import { RouteLocationRaw } from 'vue-router'
import { SpaceResource, Resource } from '@ownclouders/web-client'
import ResourceTile from '../../../../src/components/FilesList/ResourceTile.vue'

type TypeIconSize = 'large' | 'xlarge' | 'xxlarge' | 'xxxlarge'
interface Props {
  resource: SpaceResource | Resource
  resourceRoute?: RouteLocationRaw | null
  isResourceSelected?: boolean
  isResourceClickable?: boolean
  isResourceDisabled?: boolean
  isExtensionDisplayed?: boolean
  resourceIconSize?: TypeIconSize
  lazy?: boolean
}
const getSpaceMock = (props = {}) => ({
  id: 'lorem-id',
  name: 'Space 1',
  path: '',
  type: 'space',
  isFolder: true,
  disabled: false,
  getDriveAliasAndItem: () => '1',
  ...props
})

describe('OcTile component', () => {
  it('renders default space correctly', () => {
    const wrapper = getWrapper({ resource: getSpaceMock() })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders disabled space correctly', () => {
    const wrapper = getWrapper({
      resource: getSpaceMock({ disabled: true }),
      isResourceDisabled: true
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders selected resource correctly', () => {
    const wrapper = getWrapper({ resource: getSpaceMock(), isResourceSelected: true })
    expect(wrapper.find('.oc-tile-card-selected').exists()).toBeTruthy()
  })
  it('should emit click event when resource is clicked', () => {
    const wrapper = getWrapper({ resource: getSpaceMock(), isResourceSelected: true })
    const resourceLink = wrapper.find('resource-link-stub')
    resourceLink.trigger('click')
    expect(wrapper.emitted()).toHaveProperty('click')
  })
  it('should load lazily and show shimmering tile cards', () => {
    const wrapper = getWrapper({ resource: getSpaceMock(), isResourceSelected: false, lazy: true })
    expect(wrapper.find('.oc-tile-card-lazy-shimmer').exists()).toBeTruthy()
  })
  it('should show locked resource', () => {
    const wrapper = getWrapper({
      resource: getSpaceMock({ locked: true }),
      isResourceSelected: true
    })
    const element = wrapper.find('.oc-tile-card-preview')
    expect(element.attributes('aria-label')).toEqual('This item is locked')
  })
  it.each(['xlarge, xxlarge, xxxlarge'])(
    'renders resource icon size correctly',
    (resourceIconSize) => {
      const wrapper = getWrapper({
        resource: getSpaceMock(),
        resourceIconSize: resourceIconSize as TypeIconSize
      })
      expect(wrapper.find('resource-icon-stub').attributes().size).toEqual(resourceIconSize)
    }
  )

  function getWrapper(props: Props) {
    return shallowMount(ResourceTile, {
      props,
      global: { plugins: [...defaultPlugins()], renderStubDefaultSlot: true }
    })
  }
})
