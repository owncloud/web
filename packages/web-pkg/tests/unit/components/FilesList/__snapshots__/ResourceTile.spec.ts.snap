// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OcTile component > renders default space correctly 1`] = `
"<div class="oc-tile-card oc-card oc-card-default oc-rounded">
  <resource-link-stub resource="[object Object]" isresourceclickable="true" class="oc-card-media-top oc-flex oc-flex-center oc-flex-middle oc-m-rm" tabindex="-1">
    <div class="oc-tile-card-selection"></div>
    <!--v-if-->
    <div class="oc-tile-card-preview oc-flex oc-flex-middle oc-flex-center">
      <div class="oc-tile-card-hover"></div>
      <resource-icon-stub resource="[object Object]" size="xlarge" class="tile-default-image oc-pt-xs"></resource-icon-stub>
    </div>
  </resource-link-stub>
  <div class="oc-card-body oc-p-s">
    <div class="oc-flex oc-flex-between oc-flex-middle">
      <div class="oc-flex oc-flex-middle oc-text-truncate resource-name-wrapper">
        <resource-list-item-stub resource="[object Object]" pathprefix="" ispathdisplayed="false" parentfoldername="" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="true" isicondisplayed="false" isresourceclickable="true"></resource-list-item-stub>
      </div>
      <div class="oc-flex oc-flex-middle">
        <!-- Slot for indicators !-->
        <!-- Slot for individual actions -->
        <!-- Slot for contextmenu -->
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`OcTile component > renders disabled space correctly 1`] = `
"<div class="oc-tile-card oc-card oc-card-default oc-rounded oc-tile-card-disabled">
  <resource-link-stub resource="[object Object]" isresourceclickable="true" class="oc-card-media-top oc-flex oc-flex-center oc-flex-middle oc-m-rm" tabindex="-1">
    <div class="oc-tile-card-selection"></div>
    <!--v-if-->
    <div class="oc-tile-card-preview oc-flex oc-flex-middle oc-flex-center">
      <div class="oc-tile-card-hover"></div>
      <resource-icon-stub resource="[object Object]" size="xlarge" class="tile-default-image oc-pt-xs"></resource-icon-stub>
    </div>
  </resource-link-stub>
  <div class="oc-card-body oc-p-s">
    <div class="oc-flex oc-flex-between oc-flex-middle">
      <div class="oc-flex oc-flex-middle oc-text-truncate resource-name-wrapper">
        <resource-list-item-stub resource="[object Object]" pathprefix="" ispathdisplayed="false" parentfoldername="" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="true" isicondisplayed="false" isresourceclickable="true"></resource-list-item-stub>
      </div>
      <div class="oc-flex oc-flex-middle">
        <!-- Slot for indicators !-->
        <!-- Slot for individual actions -->
        <!-- Slot for contextmenu -->
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
